front-end/src/components/analysis-page/IntentMenu.jsx

```diff
@@ -0,0 +1,104 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import store from "../../store.js";
+import { setDisplayingQuestion } from '../../store/analyzeTranscriptSlice.js';
+
+class IntentMenu extends Component {
+    
+    constructor(props) {
+        super(props);
+        this.ref = React.createRef();
+        this.handleClickOutside = this.handleClickOutside.bind(this);
+        this.handleChange = this.handleChange.bind(this);
+        this.state = {
+            inputValue: "",
+        };
+    }
+
+    componentDidMount() {
+        document.addEventListener('mousedown', this.handleClickOutside, false);
+    }
+    
+    componentWillUnmount() {
+        document.removeEventListener('mousedown', this.handleClickOutside, false);
+    }
+    
+    handleClickOutside(event) {
+        // Detect when the user clicks outside of the intent menu to hide it
+        if (this.ref.current && !this.ref.current.contains(event.target)) {
+          this.props.onClickOutside();
+        }
+    }
+
+    handleChange(event) {
+        // Update the input value state based on input field
+        this.setState({inputValue: event.target.value});
+    }
+
+    listIntents = () => {
+        // Generate a list of navigation buttons with each corresponding to 1 intent from the current transcript, in order
+        return this.props.intents.map((intent, buttonIndex) => {
+
+            const goToIntent = () => {
+                // Set the transcript index to display this intent
+                const displayIndex = buttonIndex - (buttonIndex % 3);
+                store.dispatch(setDisplayingQuestion(displayIndex));
+            }
+
+            return (
+                // SEARCH FEATURE: If the user's search input matches this intent's question, then
+                intent.question.toLowerCase().includes(this.state.inputValue.toLowerCase())
+                // Generate and display a unique button element for this intent
+                ? <li className="relative" key={buttonIndex}>
+                    <button
+                        className="w-full text-left text-lg py-0 px-4 h-10 rounded truncate overflow-hidden box-border 
+                        hover:bg-gray-100 hover:text-black transition ease-in-out
+                        focus-within:bg-gray-100 focus-within:text-black"
+                        onClick={goToIntent}
+                    >
+                        {intent.question}
+                    </button>
+                </li>
+                // Otherwise, do not include this intent in the menu
+                : null
+            );
```

Comment:
IMO this can be done more cleanly. We can do the check outside of the return statement instead of using a ternary. Or even better, we can use array.filter (this.props.intents.filter().map() (maybe extract a variable)

Reply:
Ok so this actually is a question I had. Is each question in the list of intents unique? If so, then we can easily filter the array and then look up the index in the original of whichever question is clicked on in the filtered list.

The way it's being done right now is that the map loop needs to iterate thru all of the intents in the original list so that if we filter some out, the ones that come after are still at the proper index. (otherwise the indices of the filtered list do not correspond to the index needed to get there when clicking).

Reply:
Each question should be unique